### ISSUE TITLE
Uncaught 500 error: java.util.NoSuchElementException

### ISSUE THREAD

**Issue by jcoyne** [fcrepo4/fcrepo4#929]
When I do a get to: http://127.0.0.1:8080/fedora/rest/prod/5q/47/rp/18/5q47rp18c/files/3eb32fa4-ceb1-4592-9d11-a3d831b3a011/fcr:metadata

on fcrepo 4.4.0

I get
```
java.util.NoSuchElementException
	at com.google.common.collect.Iterators$1.next(Iterators.java:80)
	at org.fcrepo.kernel.modeshape.rdf.impl.mappings.PropertyValueIterator.computeNext(PropertyValueIterator.java:68)
	at org.fcrepo.kernel.modeshape.rdf.impl.mappings.PropertyValueIterator.computeNext(PropertyValueIterator.java:35)
	at com.google.common.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:143)
	at com.google.common.collect.AbstractIterator.hasNext(AbstractIterator.java:138)
	at com.google.common.collect.TransformedIterator.hasNext(TransformedIterator.java:43)
	at com.google.common.collect.Iterators$7.computeNext(Iterators.java:650)
	at com.google.common.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:143)
	at com.google.common.collect.AbstractIterator.hasNext(AbstractIterator.java:138)
	at com.google.common.collect.TransformedIterator.hasNext(TransformedIterator.java:43)
	at com.google.common.collect.Iterators$5.hasNext(Iterators.java:547)
	at com.google.common.collect.Iterators$5.hasNext(Iterators.java:547)
	at com.google.common.collect.ForwardingIterator.hasNext(ForwardingIterator.java:43)
	at com.google.common.collect.Iterators$5.hasNext(Iterators.java:547)
	at com.google.common.collect.ForwardingIterator.hasNext(ForwardingIterator.java:43)
	at com.google.common.collect.Iterators$7.computeNext(Iterators.java:650)
	at com.google.common.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:143)
	at com.google.common.collect.AbstractIterator.hasNext(AbstractIterator.java:138)
	at com.google.common.collect.Iterators$5.hasNext(Iterators.java:547)
	at com.google.common.collect.Iterators$5.hasNext(Iterators.java:547)
	at com.google.common.collect.ForwardingIterator.hasNext(ForwardingIterator.java:43)
	at com.google.common.collect.Iterators$5.hasNext(Iterators.java:547)
	at org.fcrepo.kernel.api.utils.iterators.RdfStream.asModel(RdfStream.java:303)
	at org.fcrepo.http.api.responses.StreamingBaseHtmlProvider.writeTo(StreamingBaseHtmlProvider.java:203)
	at org.fcrepo.http.api.responses.StreamingBaseHtmlProvider.writeTo(StreamingBaseHtmlProvider.java:85)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$TerminalWriterInterceptor.invokeWriteTo(WriterInterceptorExecutor.java:265)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$TerminalWriterInterceptor.aroundWriteTo(WriterInterceptorExecutor.java:250)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:162)
	at org.glassfish.jersey.server.internal.JsonWithPaddingInterceptor.aroundWriteTo(JsonWithPaddingInterceptor.java:106)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:162)
	at org.glassfish.jersey.server.internal.MappableExceptionWrapperInterceptor.aroundWriteTo(MappableExceptionWrapperInterceptor.java:85)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:162)
	at org.glassfish.jersey.message.internal.MessageBodyFactory.writeTo(MessageBodyFactory.java:1154)
	at org.glassfish.jersey.server.ServerRuntime$Responder.writeResponse(ServerRuntime.java:621)
	at org.glassfish.jersey.server.ServerRuntime$Responder.processResponse(ServerRuntime.java:377)
	at org.glassfish.jersey.server.ServerRuntime$Responder.process(ServerRuntime.java:367)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:274)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:297)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:254)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1030)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:373)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:381)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:344)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:221)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:501)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:171)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:102)
	at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:950)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:408)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1040)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:607)
	at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:314)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by awoods** [comment]
Could you please provide more context around how you created the binary resource? Because when I make a GET request on a binary/fcr:metadata, I am not seeing the error described.

**Comment by jcoyne** [comment]
No idea. I can get the file here http://127.0.0.1:8080/fedora/rest/prod/5q/47/rp/18/5q47rp18c/files/3eb32fa4-ceb1-4592-9d11-a3d831b3a011 with no issues. Just can't read the metadata node

**Comment by jcoyne** [comment]
```
curl -I http://127.0.0.1:8080/fedora/rest/prod/5q/47/rp/18/5q47rp18c/files/3eb32fa4-ceb1-4592-9d11-a3d831b3a011
HTTP/1.1 200 OK
Server: Apache-Coyote/1.1
ETag: "b2da94ab40d84b027d46a01c356b3bafbb0bcba0"
Last-Modified: Thu, 29 Oct 2015 04:43:19 GMT
Accept-Ranges: bytes
Content-Disposition: attachment; filename="avi14885.00654.023.xml"; creation-date="Thu, 29 Oct 2015 04:37:40 GMT"; modification-date="Thu, 29 Oct 2015 04:43:19 GMT"; size=52041
Link: <http://www.w3.org/ns/ldp#Resource>;rel="type"
Link: <http://www.w3.org/ns/ldp#NonRDFSource>;rel="type"
Link: <http://127.0.0.1:8080/fedora/rest/prod/5q/47/rp/18/5q47rp18c/files/3eb32fa4-ceb1-4592-9d11-a3d831b3a011/fcr:metadata>; rel="describedby"
Allow: DELETE,HEAD,GET,PUT,OPTIONS
Content-Type: text/xml
Content-Length: 52041
Date: Thu, 29 Oct 2015 17:50:21 GMT
```

**Comment by ojlyytinen** [comment]
I ran into this same stack trace and can reproduce it consistently. I'm using Fedora 4.5.0 through HydraWorks and ActiveFedora. Originally I encountered this when uploading new versions of a file using Hydra::Works::AddFileToFileSet. This would always work for the initial file version and for one updated version but would consistently fail when trying to update the file the second time.

I tried to produce a minimal test case and came up with this bit of code

```ruby
f=Hydra::PCDM::File.new
f.content='foo'
f.save
f.create_version
f=Hydra::PCDM::File.find(f.id)
f.save
f.create_version
f=Hydra::PCDM::File.find(f.id)
f.save
f.create_version
f=Hydra::PCDM::File.find(f.id)
f.save
```
It's always the last save which fails and produces the stack trace. The error happens inside save but according to the Ruby stack trace it's really trying to access the metadata node when the error happens.

If I remove the file versioning then it works fine. Also reloading the file between saving is essential to reproduce the error.

**Comment by awoods** [comment]
It would be helpful to know what is happening in the hydra application in the above method calls. When reproducing the steps above directly against Fedora, I am not seeing an error:
```
curl -i -XPUT -H"Content-Type: text/plain" --data-binary @file.txt localhost:8080/rest/file
curl -i -XPOST -H"slug: v0" localhost:8080/rest/file/fcr:versions
curl -i -XGET localhost:8080/rest/file
curl -i -XPUT -H"Content-Type: text/plain" --data-binary @file.txt localhost:8080/rest/file
curl -i -XPOST -H"slug: v1" localhost:8080/rest/file/fcr:versions
curl -i -XGET localhost:8080/rest/file
curl -i -XPUT -H"Content-Type: text/plain" --data-binary @file.txt localhost:8080/rest/file
curl -i -XPOST -H"slug: v2" localhost:8080/rest/file/fcr:versions
curl -i -XGET localhost:8080/rest/file
curl -i -XPUT -H"Content-Type: text/plain" --data-binary @file.txt localhost:8080/rest/file
```

**Comment by ojlyytinen** [comment]
I captured the http traffic between Hydra and Fedora and I think I got the essential bits. This series http requests causes the error to happen consistently, at least on my machine.

```
curl -i -XPOST -H"Slug: testfile" -H"Content-Type: text/plain" --data-binary @file.txt http://fedoraAdmin:fedoraAdmin@localhost:8983/fedora/rest/dev
curl -i -XPATCH -H"Content-Type: application/sparql-update" --data-binary @patch.txt http://fedoraAdmin:fedoraAdmin@localhost:8983/fedora/rest/dev/testfile/fcr:metadata
curl -i -XPOST -H"Slug: version1" http://fedoraAdmin:fedoraAdmin@localhost:8983/fedora/rest/dev/testfile/fcr:versions
curl -i -XPATCH -H"Content-Type: application/sparql-update" --data-binary @patch2.txt http://fedoraAdmin:fedoraAdmin@localhost:8983/fedora/rest/dev/testfile/fcr:metadata
curl -i -XPOST -H"Slug: version2" http://fedoraAdmin:fedoraAdmin@localhost:8983/fedora/rest/dev/testfile/fcr:versions
curl -i -XPATCH -H"Content-Type: application/sparql-update" --data-binary @patch2.txt http://fedoraAdmin:fedoraAdmin@localhost:8983/fedora/rest/dev/testfile/fcr:metadata
curl -i -XPOST -H"Slug: version3" http://fedoraAdmin:fedoraAdmin@localhost:8983/fedora/rest/dev/testfile/fcr:versions
curl -i -XGET http://fedoraAdmin:fedoraAdmin@localhost:8983/fedora/rest/dev/testfile/fcr:metadata
```

file.txt is just any plain text file. patch.txt is
```
DELETE { <http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> ?change . }
  WHERE { <http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> ?change . } ;
INSERT { 
<http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/ldp#NonRDFSource> .
<http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.jcp.org/jcr/nt/1.0resource> .
<http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.jcp.org/jcr/mix/1.0mimeType> .
<http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://fedora.info/definitions/v4/repository#Binary> .
<http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://fedora.info/definitions/v4/repository#Resource> .
<http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://pcdm.org/models#File> .
}
 WHERE { }
```

And patch2.txt is
```
DELETE { <http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> ?change . }
  WHERE { <http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> ?change . } ;
INSERT { 
<http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/ldp#NonRDFSource> .
<http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.jcp.org/jcr/nt/1.0resource> .
<http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.jcp.org/jcr/mix/1.0mimeType> .
<http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://fedora.info/definitions/v4/repository#Binary> .
<http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://fedora.info/definitions/v4/repository#Resource> .
<http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://pcdm.org/models#File> .
<http://localhost:8983/fedora/rest/dev/testfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.jcp.org/jcr/mix/1.0versionable> .
}
 WHERE { }
```

**Comment by escowles** [comment]
I see this error when I run this script against 4.5.0, but I do not see it when running against a current 4.5.1-RC build.  So I think this error has been fixed and will be resolved by the upcoming 4.5.1 release.

**Comment by ojlyytinen** [comment]
I just tried in 4.5.1-RC3 as well and the problem seems to have been fixed. Updating files several times through Hydra is also working fine.

Possibly fixed by 523953db47ff09c0311de1dd51e98ae384cc0836? That commit seems to have thoroughly changed the iterators where the error occurred. Although I couldn't say if the root cause was in the iterators or somewhere else.

**Comment by acoburn** [comment]
This was fixed by much of the java8 refactoring of the last few months. 

This is the relevant commit: https://github.com/fcrepo4/fcrepo4/commit/f0db4f21a1dcf72eeb2ffb565462ab7fc4600a8c

and this was the line in question: https://git.io/vwWLQ

Monads FTW!
